ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.ts

```diff
@@ -21,47 +22,79 @@ export class FileDownloadComponent implements OnDestroy {
   downloading = false;
   CLEAN = 'CLEAN';
   subscription: Subscription;
+  subscription2: Subscription;
 
-  constructor(private dsmService: DSMService ) {
+  constructor( private dsmService: DSMService ) {
 
   }
 
   ngOnDestroy(): void {
     if (this.subscription != null) {
       this.subscription.unsubscribe();
     }
+    if (this.subscription2 != null) {
+      this.subscription2.unsubscribe();
+    }
   }
 
   downloadParticipantFile( file: File ): void {
     this.downloading = true;
-    if (file.scanResult !== this.CLEAN) {
+    if (!this.isFileClean( file )) {
       this.downloadMessage = 'Error - file has not passed scanning';
       this.downloading = false;
       return;
     }
     const realm = localStorage.getItem( ComponentService.MENU_SELECTED_REALM );
-    this.subscription = this.dsmService.downloadParticipantFile( file.fileName, file.bucket, file.blobName, realm, file.mimeType )
-      .subscribe( {
+    this.subscription = this.dsmService.getSignedUrl( this.participant.data.profile[ 'guid' ], file.fileName, file.bucket,
+      file.blobName, file.guid, realm ).subscribe( {
         next: data => {
-          this.saveParticipantFile( data, file.mimeType, file.fileName );
+          this.saveParticipantFile( data.url, file.mimeType, data.fileName );
+        },
```

Comment:
When using rxjs it is generally considered bad practice to call another observable from the `.subscribe` handler.
Also, in current code you are creating subscriptions each time the method `downloadParticipantFile` is being called.

You are assigning the subscription to a variable to later close (good!), but every time the method gets called, you create a new subscription and the previous subscription goes unassigned and unclosed.


A single pipeline is going to be cleaner simpler and you wont even have to worry about closing subscriptions.
So you want to do something like:
```typescript
this.dsmService.getSignedUrl( this.participant.data.profile[ 'guid' ], etc.etc. ).pipe(
   .mergeMap(data -> downloadFromSignedUrl(data.url),
   .take(1))
   .subscribe({
        next: data => {
          const blob = new Blob( [ data ], {type: type} );
          fileSaver.saveAs( blob, fileName );
          this.downloading = false;
          this.downloadMessage = 'File download finished.';
        },
        error: err => {
          console.log( err );
          this.downloadMessage = err;
          this.downloading = false;
        }
      }
```
The take(1) will unsubscribe every time it gets called. You only need to the error handler once etc.
Try it and see if it flies.

Reply:
Agree, that would be much better.

Reply:
Thanks Marco! This was a great tip! It is fixed now